File: migration/backend/app/services/yahoo.py
"""Yahoo chart client — the server half of the Stack page's live quotes.

Ported from the Cloudflare Worker that used to do this job
(automation/worker/live-quotes.js in the old schwarzrene.github.io repo).

Yahoo's chart endpoint sends no CORS headers, so a browser cannot call it
directly. This module calls it server-side and returns the same column-wise
shape the committed snapshot uses, so a level drawn on one sits at the same
height on the other.

Prices are adjusted by the adjclose ratio to match the snapshot written by
automation/scripts/fetch_market_data.py, which uses yfinance auto_adjust=True.
"""

from __future__ import annotations

import asyncio
import logging
import re
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any

import httpx

log = logging.getLogger(__name__)

# One to ten characters: letters, digits, dot, hyphen. Anything else is not a
# ticker and is rejected before it reaches Yahoo.
SYMBOL_RE = re.compile(r"^[A-Z0-9][A-Z0-9.\-]{0,9}$")

USER_AGENT = (
    "Mozilla/5.0 (compatible; bqe-backend/1.0; +https://github.com/SchwarzRene/bqe-backend)"
)


@dataclass(frozen=True)
class Series:
    """One Yahoo range, described the way the page wants to read it."""

    range: str
    interval: str
    unit_ms: int  # multiplier that turns a stored `t` value back into milliseconds


DAILY = Series(range="10y", interval="1d", unit_ms=86_400_000)
HOURLY = Series(range="60d", interval="60m", unit_ms=60_000)

# Below this many daily bars the card has nothing worth drawing.
MIN_DAILY_BARS = 40
MIN_HOURLY_BARS = 20


class UpstreamError(RuntimeError):
    """Yahoo refused, timed out, or answered with something unusable."""


def is_valid_symbol(symbol: str) -> bool:
    return bool(SYMBOL_RE.match(symbol))


async def fetch_quote(client: httpx.AsyncClient, symbol: str, base_url: str) -> dict[str, Any]:
    """Daily bars, plus hourly when Yahoo offers them.

    The hourly series is a bonus: if it fails the daily series is still
    returned, because the chart is drawn from the daily bars.
    """
    daily_task = _chart(client, symbol, DAILY, base_url)
    hourly_task = _chart(client, symbol, HOURLY, base_url)

    daily, hourly = await asyncio.gather(daily_task, hourly_task, return_exceptions=True)

    if isinstance(daily, BaseException):
        raise UpstreamError(str(daily)) from daily
    if daily is None or len(daily["t"]) < MIN_DAILY_BARS:
        raise UpstreamError("no data")

    payload: dict[str, Any] = {
        "s": symbol,
        "updated": datetime.now(UTC).isoformat(timespec="seconds").replace("+00:00", "Z"),
        "live": True,
        "d": daily,
    }

    if isinstance(hourly, BaseException):
        log.info("hourly series unavailable for %s: %s", symbol, hourly)
    elif hourly is not None and len(hourly["t"]) > MIN_HOURLY_BARS:
        payload["h1"] = hourly

    return payload


async def _chart(
    client: httpx.AsyncClient, symbol: str, spec: Series, base_url: str
) -> dict[str, Any] | None:
    """One Yahoo range, returned in the page's column format."""
    url = f"{base_url.rstrip('/')}/v8/finance/chart/{symbol}"
    params = {"range": spec.range, "interval": spec.interval, "includePrePost": "false"}

    response = await client.get(url, params=params, headers={"User-Agent": USER_AGENT})
    if response.status_code != 200:
        raise UpstreamError(f"yahoo {response.status_code}")

    body = response.json()
    result = (((body or {}).get("chart") or {}).get("result") or [None])[0]
    if not result or not isinstance(result.get("timestamp"), list):
        return None

    indicators = result.get("indicators") or {}
    quote = (indicators.get("quote") or [{}])[0] or {}
    adjclose = ((indicators.get("adjclose") or [{}])[0] or {}).get("adjclose") or []

    opens, highs, lows, closes = (
        quote.get("open") or [],
        quote.get("high") or [],
        quote.get("low") or [],
        quote.get("close") or [],
    )

    step = spec.unit_ms // 1000
    out: dict[str, Any] = {"tu": spec.unit_ms, "t": [], "o": [], "h": [], "l": [], "c": []}

    for i, stamp in enumerate(result["timestamp"]):
        o, h, low, c = (
            _at(opens, i),
            _at(highs, i),
            _at(lows, i),
            _at(closes, i),
        )
        if any(v is None or not v > 0 for v in (o, h, low, c)):
            continue

        # Match the snapshot's auto_adjust=True: scale OHLC by adjclose/close.
        adj = _at(adjclose, i)
        k = adj / c if adj is not None and c > 0 else 1.0
        digits = 4 if c * k < 10 else 2

        out["t"].append(stamp // step)
        out["o"].append(round(o * k, digits))
        out["h"].append(round(h * k, digits))
        out["l"].append(round(low * k, digits))
        out["c"].append(round(c * k, digits))

    return out


def _at(values: list[Any], index: int) -> float | None:
    """Yahoo pads its arrays with nulls and sometimes ends them early."""
    if index >= len(values):
        return None
    value = values[index]
    return None if value is None else float(value)

File: worker/yahoo.ts
// Yahoo chart client. Ported from bqe-backend's app/services/yahoo.py, which
// was itself a port of the Worker this site used before that.
//
// Yahoo's chart endpoint sends no CORS headers, so the browser cannot call it
// directly; the Worker does, and returns the same column-wise shape the Stack
// page reads. Prices are scaled by adjclose/close, matching yfinance's
// auto_adjust=True, so a level drawn on live bars sits at the same height on
// the stored ones.

const BASE = "https://query1.finance.yahoo.com";
const USER_AGENT = "Mozilla/5.0 (compatible; bqe-frontend/1.0; +https://github.com/SchwarzRene/bqe-frontend)";

// One to ten characters: letters, digits, dot, hyphen. Anything else is not
// a ticker and never reaches Yahoo.
const SYMBOL_RE = /^[A-Z0-9][A-Z0-9.\-]{0,9}$/;

export interface Columns {
  tu: number; // multiplier that turns a stored `t` back into milliseconds
  t: number[];
  o: number[];
  h: number[];
  l: number[];
  c: number[];
}

interface Spec {
  range: string;
  interval: string;
  unitMs: number;
}

export const DAILY: Spec = { range: "10y", interval: "1d", unitMs: 86_400_000 };
export const HOURLY: Spec = { range: "60d", interval: "60m", unitMs: 60_000 };

// Below this many daily bars a card has nothing worth drawing.
const MIN_DAILY_BARS = 40;
const MIN_HOURLY_BARS = 20;

export class UpstreamError extends Error {}

export function isValidSymbol(symbol: string): boolean {
  return SYMBOL_RE.test(symbol);
}

/** BRK.B -> BRK-B: Yahoo's spelling, and the file name the page asks for. */
export function yahooSymbol(symbol: string): string {
  return symbol.trim().toUpperCase().replace(/\./g, "-");
}

export interface Quote {
  d: Columns;
  h1?: Columns;
}

/**
 * Daily bars, plus hourly when Yahoo has them. The hourly series is a bonus:
 * if it fails the daily one is still returned, since the chart is drawn from
 * the daily bars.
 */
export async function fetchQuote(symbol: string, fetcher: typeof fetch = fetch): Promise<Quote> {
  const [daily, hourly] = await Promise.allSettled([
    chart(symbol, DAILY, fetcher),
    chart(symbol, HOURLY, fetcher),
  ]);
  if (daily.status === "rejected") {
    throw daily.reason instanceof UpstreamError ? daily.reason : new UpstreamError(String(daily.reason));
  }
  if (!daily.value || daily.value.t.length < MIN_DAILY_BARS) throw new UpstreamError("no data");

  const quote: Quote = { d: daily.value };
  if (hourly.status === "fulfilled" && hourly.value && hourly.value.t.length > MIN_HOURLY_BARS) {
    quote.h1 = hourly.value;
  }
  return quote;
}

async function chart(symbol: string, spec: Spec, fetcher: typeof fetch): Promise<Columns | null> {
  const url = new URL(`/v8/finance/chart/${encodeURIComponent(symbol)}`, BASE);
  url.searchParams.set("range", spec.range);
  url.searchParams.set("interval", spec.interval);
  url.searchParams.set("includePrePost", "false");

  const res = await fetcher(url.toString(), {
    headers: { "User-Agent": USER_AGENT, Accept: "application/json" },
    signal: AbortSignal.timeout(8000),
  });
  if (res.status !== 200) throw new UpstreamError(`yahoo ${res.status}`);
  return toColumns(await res.json(), spec.unitMs);
}

/** A Yahoo chart response in the page's column format. Exported for tests. */
export function toColumns(body: any, unitMs: number): Columns | null {
  const result = body?.chart?.result?.[0];
  if (!result || !Array.isArray(result.timestamp)) return null;

  const quote = result.indicators?.quote?.[0] ?? {};
  const adjclose: (number | null)[] = result.indicators?.adjclose?.[0]?.adjclose ?? [];
  const opens = quote.open ?? [], highs = quote.high ?? [], lows = quote.low ?? [], closes = quote.close ?? [];

  const step = unitMs / 1000;
  const out: Columns = { tu: unitMs, t: [], o: [], h: [], l: [], c: [] };

  result.timestamp.forEach((stamp: number, i: number) => {
    const o = opens[i], h = highs[i], l = lows[i], c = closes[i];
    if (![o, h, l, c].every((v) => typeof v === "number" && v > 0)) return;

    // Match the snapshot's auto_adjust=True: scale OHLC by adjclose/close.
    const adj = adjclose[i];
    const k = typeof adj === "number" && adj > 0 ? adj / c : 1;
    const digits = c * k < 10 ? 4 : 2;

    out.t.push(Math.floor(stamp / step));
    out.o.push(round(o * k, digits));
    out.h.push(round(h * k, digits));
    out.l.push(round(l * k, digits));
    out.c.push(round(c * k, digits));
  });
  return out;
}

function round(value: number, digits: number): number {
  const f = 10 ** digits;
  return Math.round(value * f) / f;
}
